### PR TITLE
IT3: Full simulation + Glo-tracking

### PR DESCRIPTION
**Pull request by f3sch**
This PR aims to incorporate IT3 into the global-reconstruction chain + make simulation a bit easier:
 - Finds IT3-TPC(-TRD,-TOF) matches, the relevant workflows have now a 'GlobalParams.withITS3' option
 - Reflects the necessary changes in IT3 code

The flags and functionality is hidden behind the compile flag 'ENABLE_UPGRADES' + the default is off.
I tested the full simulation chain for ENABLE_UPGRADES=On/Off + w. & w/o. IT3. 

I can also make different PRs (however reviewers want :)). A brief overview:
 - TRD: 4 Files (since the matching uses ITS clusters, need to optionally load different topology dictionary)
 - ITS: 3 Files (change calculation for normal of IT3 + change printf to LOGF when printing the RecoGeomHelper)
 - ITS-TPC: 7 Files, getting matching to work
 - Rest is IT3 only changes

### CODE REVIEW

**Comment by mconcas** [comment]
Hi @f3sch,
From the ITS perspective, there are little to no changes, so to me, it is okay.

I have a couple of more general comments/questions:

1) The "`IT3`" is a nickname to cope with FairRoot's inability to message something with more than three characters in the source ID. In the case of sensitive detectors in simulations, it is the `DetID`. I don't know all the details, but this is why we have this arguably ugly `IT3` name.
One could also consider exposing the extended `ITS3` name to the user (e.g., in flags, branch names, etc.). I don't mind the naming, but we should ensure this is generally okay since it will quickly crystallize in scripts/automation, etc. I think this is not doable in the simulation command, as the detector names are then parsed as DetID stringified names somewhere.

2) Are you sure you don't want to make "ITS3 -> ITS spoofing" optional for readers and writers? Maybe even with inverse logic, with spoofing being the default and the possibility of producing something that is ITS3-explicit? This is more of a curiosity. I know this PR is targeting global reco, which requires this trick.

3) I do personally like the `PUBLIC_UPGRADE_LINK_LIBRARIES` extension (should not `PUBLIC_LINK_LIBRARIES_UPGRADE` sound better?). Now, if this is ok to have (pinging @davidrohr and @ktf for more authoritative comments), I think this should be extended to all of the CMake code that could link against libraries for upgrades, including dictionary creation and executables that now are living with explicit checks (e.g. [here]( https://github.com/AliceO2Group/AliceO2/blob/c34b2a3a61f1e1ebf39ad7c4d2c179d82e3c4ca8/macro/CMakeLists.txt#L126)) and the rest should be changed accordingly, so to make the feature globally available and uniformly used. A dedicated PR could work for this.

**Comment by f3sch** [comment]
Hi @mconcas, 

Thanks for the feedback!

> 1. The "`IT3`" is a nickname to cope with FairRoot's inability to message something with more than three characters in the source ID. In the case of sensitive detectors in simulations, it is the `DetID`. I don't know all the details, but this is why we have this arguably ugly `IT3` name.

Agreed.

>    One could also consider exposing the extended `ITS3` name to the user (e.g., in flags, branch names, etc.). I don't mind the naming, but we should ensure this is generally okay since it will quickly crystallize in scripts/automation, etc. I think this is not doable in the simulation command, as the detector names are then parsed as DetID stringified names somewhere.

One example for this would be in the digit branch names where the detID is stringified. Also, I have no preference on how the flags for the workflows are named, so I will change them.

> 2. Are you sure you don't want to make "ITS3 -> ITS spoofing" optional for readers and writers? Maybe even with inverse logic, with spoofing being the default and the possibility of producing something that is ITS3-explicit? This is more of a curiosity. I know this PR is targeting global reco, which requires this trick.

I do not mind implementing this switch but what would be the use-case of having it, since ITS and ITS3 are mutually exclusive so there will not be any overlap in the first place?

> 3. I do personally like the `PUBLIC_UPGRADE_LINK_LIBRARIES` extension (should not `PUBLIC_LINK_LIBRARIES_UPGRADE` sound better?). Now, if this is ok to have (pinging @davidrohr and @ktf for more authoritative comments), I think this should be extended to all of the CMake code that could link against libraries for upgrades, including dictionary creation and executables that now are living with explicit checks (e.g. [here](https://github.com/AliceO2Group/AliceO2/blob/c34b2a3a61f1e1ebf39ad7c4d2c179d82e3c4ca8/macro/CMakeLists.txt#L126)) and the rest should be changed accordingly, so to make the feature globally available and uniformly used. A dedicated PR could work for this.

Naming things is hard.... and yes, your suggestions makes more sense, will change.

**Comment by f3sch** [comment]
Failures seem unrelated.

@sawenzel When you have time, could you please take a look at the proposed changes for `o2-sim` part?

**Comment by f3sch** [comment]
@ktf When you have time, would you please take a look at the proposed CMake changes? Thanks.

**Comment by ktf** [comment]
Why don't you use an expression for the upgrade libraries? I do not like the idea of embedding too much extensions to o2_add_library. In general that's a recipe for disasters.

**Comment by ktf** [comment]
I.e.  `$<$<BOOL:${ENABLE_UPGRADE}>:optional_library>`

**Comment by sawenzel** [comment]
@f3sch @mconcas . In general nice developments and big progress! But I have a few comments to this PR.

(a) just f.y.i; as far as I remember the limit of 3 letters for DetID strings does not come from FairRoot! I believe this is more part of the DPL (header) protocol.

(b) I am not a big fan of adding new specific options to `o2-sim` just to modify a detector list in the sense of `o2-sim --withITS3` (because it doesn't scale when more detector have wishes like this). Here are my thoughts: 
  
  * First of all, it is probably not urgently needed since you had a solution by manually giving the complete list
  * Secondly, I already discussed with @mconcas an alternative path using ConfigurableParams. One could for instance already now use something like `o2-sim --configKeyValues "ITS.DetectorVersion=3"` ... and then just construct the ITS3 geometry instead of ITS2 in the build_geometry.C setup, depending on this parameter (understanding that -m ITS would stand as proxy for either versions).
  * Overall, I agree that we need better flexibility in constructing detector lists for o2-sim. This is a already a longer term planned development. But I would like to make it more general by doing something like `o2-sim --detector-version [ALICE2 | ALICE2.1 | ALICE3 | CUSTOM]` and then potentially define the list of detectors for these names in some json or other file.

(c) Along the same line, is it really necessary to modify multiple function signatures with new arguments `bool withIT3`. Again an alternative might be just using configKeyParams... but since this concerns more the reco part maybe @shahor02 has different preferences/opinions.

**Comment by f3sch** [comment]
Thanks @ktf, dropped the cmake changes and replaced them by expressions.

**Comment by f3sch** [comment]
@sawenzel thanks for the comments, dropped the sim changes in favour of the other PR and indeed using the configKeyValues is better.

**Comment by f3sch** [comment]
Hi @wille10, the changes for TRD should be minimal. I added a configKeyValue as for TRD there wasn't one before, for now this only holds a single value but maybe in the future can be extended. Do the changes seem fine to you?

**Comment by martenole** [comment]
> Hi @wille10, the changes for TRD should be minimal. I added a configKeyValue as for TRD there wasn't one before, for now this only holds a single value but maybe in the future can be extended. Do the changes seem fine to you?

For the TRD reconstruction the parameters are stored in `GPU/GPUTracking/Definitions/GPUSettingsList.h`. Why is a new parameter class needed? Can it not be a wf option?

**Comment by f3sch** [comment]
> For the TRD reconstruction the parameters are stored in `GPU/GPUTracking/Definitions/GPUSettingsList.h`. Why is a new parameter class needed? Can it not be a wf option?

@martenole sorry I pinged the wrong person. I would prefer to have the settings consistent across the workflows, e.g., using parameters and not wf options. Correct me if I am wrong, the GPUSettings seem to be loaded after the request for the ITS clusters has been made, for ITS3 we need to conditionally load a different topology dictionary hence I need it to the input spec.

**Comment by martenole** [comment]
> > For the TRD reconstruction the parameters are stored in `GPU/GPUTracking/Definitions/GPUSettingsList.h`. Why is a new parameter class needed? Can it not be a wf option?
> 
> @martenole sorry I pinged the wrong person. I would prefer to have the settings consistent across the workflows, e.g., using parameters and not wf options. Correct me if I am wrong, the GPUSettings seem to be loaded after the request for the ITS clusters has been made, for ITS3 we need to conditionally load a different topology dictionary hence I need it to the input spec.

Hm, ok. The GPU settings are indeed not the place for that setting. And the configurable parameter classes that we have (for calibration and simulation) should not be abused for this. Then I think it is OK to add the reconstruction class. And no worries, @wille10 is the correct person to tag for TRD now, but I also saw it so thought I can as well reply :)

**Comment by shahor02** [comment]
@f3sch sorry if I missed something, but why not defining a single key `withITS3` rather than doing this in different params?
You can even add it as a member of the NameConf (in the same lib. as ConfigurableParam) if some paths should be generated accounting for the ITS3.
BTW, the only thing which prevents initializing configurableParam from the workflow (instead of the device spec builder) is that currently the global (workflow) and local (device) options cannot be the same, so we cannot put `--configKeyValues` on both levels. But this, I guess, is fixable. Obviously, in this case the configurable cannot be initialized from the ccdb before the 1st TF.

**Comment by f3sch** [comment]
> Hm, ok. The GPU settings are indeed not the place for that setting. And the configurable parameter classes that we have (for calibration and simulation) should not be abused for this. Then I think it is OK to add the reconstruction class. And no worries, @wille10 is the correct person to tag for TRD now, but I also saw it so thought I can as well reply :)

Thanks for taking the time :)

**Comment by f3sch** [comment]
> @f3sch sorry if I missed something, but why not defining a single key `withITS3` rather than doing this in different params? You can even add it as a member of the NameConf (in the same lib. as ConfigurableParam) if some paths should be generated accounting for the ITS3. BTW, the only thing which prevents initializing configurableParam from the workflow (instead of the device spec builder) is that currently the global (workflow) and local (device) options cannot be the same, so we cannot put `--configKeyValues` on both levels. But this, I guess, is fixable. Obviously, in this case the configurable cannot be initialized from the ccdb before the 1st TF.

Hi @shahor02, indeed it might be nicer to have it more as a global option instead of scattering this about in different files. I took a look but only found `KeyVal.cxx` which seems like one, but this file, AFAICT, defines the input, output dirs for the configparams so it might not be appropriate. Do you have another one mind? For the NameConf, I don't think we need to generate different paths right now.

**Comment by shahor02** [comment]
@f3sch then you can add a new ConfigurableParam, e.g. GlobalParams in the Detectors/Base, could be useful also for the future additions.

**Comment by f3sch** [comment]
@shahor02 thanks, removed all other additions to configparams in favour of newly added GlobalParms.

**Comment by f3sch** [comment]
Hi @shahor02, when you have time could you please take a look if you're fine with the proposed changes for the matching?

**Comment by shahor02** [comment]
@f3sch could you rebase it to dev? I see that many modifications are already in the dev.

**Comment by f3sch** [comment]
Hi @shahor02, I rebased but did not see any conflicts, which modifications did you spot that were already upstreamed?

File: DataFormats/Detectors/GlobalTracking/include/DataFormatsGlobalTracking/RecoContainer.h [hunk 1]
```diff
@@ -247,6 +247,10 @@ struct DataRequest {
   void requestStrangeTracks(bool mc);
 
   void requestIRFramesITS();
+
+#ifdef ENABLE_UPGRADES
```
Comment (shahor02) [inline DataFormats/Detectors/GlobalTracking/include/DataFormatsGlobalTracking/RecoContainer.h:251]: in principle, entries like that do not need `ENABLE_UPGRADES` guards, only those which alter current code execution or conflict with it.

File: DataFormats/Detectors/GlobalTracking/src/RecoContainer.cxx [hunk 2]
```diff
@@ -673,6 +692,13 @@ void RecoContainer::collectData(ProcessingContext& pc, const DataRequest& reques
     addITSClusters(pc, req->second);
   }
 
+#ifdef ENABLE_UPGRADES
```
Comment (shahor02) [inline DataFormats/Detectors/GlobalTracking/src/RecoContainer.cxx:695]: i.e. here it is surely needed.

File: DataFormats/Detectors/GlobalTracking/src/RecoContainer.cxx [hunk 3]
```diff
@@ -1036,8 +1062,23 @@ void RecoContainer::addHMPMatches(ProcessingContext& pc, bool mc)
...
   if (pc.services().get<o2::framework::TimingInfo>().globalRunNumberChanged) {            // this params need to be queried only once
+    pc.inputs().get<o2::itsmft::DPLAlpideParam<o2::detectors::DetID::ITS>*>("alpparITS"); // note: configurable param does not need finaliseCCDB
     pc.inputs().get<o2::itsmft::TopologyDictionary*>("cldictITS");                        // just to trigger the finaliseCCDB
+  }
+  commonPool[GTrackID::ITS].registerContainer(pc.inputs().get<gsl::span<o2::itsmft::ROFRecord>>("clusITSROF"), CLUSREFS);
+  commonPool[GTrackID::ITS].registerContainer(pc.inputs().get<gsl::span<o2::itsmft::CompClusterExt>>("clusITS"), CLUSTERS);
+  commonPool[GTrackID::ITS].registerContainer(pc.inputs().get<gsl::span<unsigned char>>("clusITSPatt"), PATTERNS);
+  if (mc) {
+    mcITSClusters = pc.inputs().get<const dataformats::MCTruthContainer<MCCompLabel>*>("clusITSMC");
+  }
```
Comment (shahor02) [inline DataFormats/Detectors/GlobalTracking/src/RecoContainer.cxx]: These lines are already in dev, though 1065/66 are swapped

**Review by shahor02 — DISMISSED** [review]
Looks fine, many `#ifdef's` are actually not necessary (though don't harm either)

**Comment by shahor02** [comment]
@f3sch do you expect more reviews or shall I merge it?

File: DataFormats/Detectors/GlobalTracking/include/DataFormatsGlobalTracking/RecoContainer.h (hunk 1 above)
Comment (f3sch) [reply DataFormats/Detectors/GlobalTracking/include/DataFormatsGlobalTracking/RecoContainer.h:251]: Swapped 1066/65 back, apparently some remnant of my changes...

**Comment by f3sch** [comment]
Hi @shahor02, sorry somehow I missed the email notifications on your review. Thanks!
If you are fine with how I *butchered* the matching code, then this would be the final review.

**Review by shahor02 — APPROVED** [review]

**Comment by alcaliva** [comment]
label async-2023-pbpb-apass4 added because this PR is needed to be able to apply others

**Pull request merged**